File: java/Check.java
// Driver interoperability check for ExecDB's pgwire (spec §8, phase 4
// Step 7): pgJDBC, used with its own default connection settings (no
// ExecDB-specific workaround flags -- notably preferQueryMode, which
// defaults to "extended", phase 4 Step 5), against a running ExecDB
// instance already seeded with table t(a INTEGER) (seeded by run-all.sh below).
//
// JDBC's default auto-commit=true means no statement here runs inside an
// implicit transaction, unlike psycopg2's default (python/
// check.py) -- so the DDL-rejection check below needs no explicit
// rollback to recover afterward.
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;

public class Check {
    public static void main(String[] args) throws Exception {
        if (args.length != 1) {
            System.err.println("usage: Check <jdbc-url>");
            System.exit(2);
        }
        try (Connection conn = DriverManager.getConnection(args[0])) {
            check(conn);
        }
        System.out.println("OK");
    }

    static void check(Connection conn) throws SQLException {
        try (PreparedStatement ps = conn.prepareStatement("SELECT ?")) {
            ps.setInt(1, 1);
            try (ResultSet rs = ps.executeQuery()) {
                rs.next();
                long one = rs.getLong(1);
                if (one != 1) {
                    throw new AssertionError("SELECT ?(1) returned " + one);
                }
            }
        }

        try (PreparedStatement ps = conn.prepareStatement("SELECT ?")) {
            ps.setDouble(1, 3.5);
            try (ResultSet rs = ps.executeQuery()) {
                rs.next();
                double f = rs.getDouble(1);
                if (f != 3.5) {
                    throw new AssertionError("SELECT ?(3.5) returned " + f);
                }
            }
        }

        try (PreparedStatement ps = conn.prepareStatement("SELECT ?")) {
            ps.setString(1, "hello");
            try (ResultSet rs = ps.executeQuery()) {
                rs.next();
                String s = rs.getString(1);
                if (!"hello".equals(s)) {
                    throw new AssertionError("SELECT ?('hello') returned " + s);
                }
            }
        }

        try (Statement st = conn.createStatement();
                ResultSet rs = st.executeQuery("SELECT x'00ff'")) {
            rs.next();
            byte[] blob = rs.getBytes(1);
            if (blob.length != 2 || blob[0] != 0x00 || (blob[1] & 0xFF) != 0xFF) {
                throw new AssertionError("SELECT x'00ff' returned an unexpected byte sequence");
            }
        }

        try (Statement st = conn.createStatement();
                ResultSet rs = st.executeQuery("SELECT NULL")) {
            rs.next();
            rs.getObject(1);
            if (!rs.wasNull()) {
                throw new AssertionError("SELECT NULL did not report wasNull()");
            }
        }

        // spec §2: DDL must be rejected via the external I/F, and pgJDBC
        // must surface it as an SQLException carrying SQLSTATE 42501
        // (spec §8), not just some generic failure.
        try (Statement st = conn.createStatement()) {
            st.execute("CREATE TABLE jdbc_should_not_exist(a INTEGER)");
            throw new AssertionError("expected CREATE TABLE to be rejected via the external I/F");
        } catch (SQLException e) {
            if (!"42501".equals(e.getSQLState())) {
                throw new AssertionError(
                        "expected SQLSTATE 42501, got " + e.getSQLState() + " (" + e.getMessage() + ")");
            }
        }

        // A basic write/read round trip against the table run-all.sh seeded.
        try (PreparedStatement ps = conn.prepareStatement("INSERT INTO t VALUES (?)")) {
            ps.setInt(1, 777003);
            ps.executeUpdate();
        }
        try (PreparedStatement ps = conn.prepareStatement("SELECT count(*) FROM t WHERE a = ?")) {
            ps.setInt(1, 777003);
            try (ResultSet rs = ps.executeQuery()) {
                rs.next();
                long n = rs.getLong(1);
                if (n != 1) {
                    throw new AssertionError("expected count=1 after INSERT, got " + n);
                }
            }
        }
    }
}

File: python/check.py
#!/usr/bin/env python3
"""Driver interoperability check for ExecDB's pgwire (spec §8, phase 4
Step 7): psycopg2, used with its own default connection settings (no
ExecDB-specific workaround flags), against a running ExecDB instance
already seeded with table t(a INTEGER) (seeded by run-all.sh below).

psycopg2 wraps every statement in an implicit BEGIN/COMMIT (autocommit is
off by default) by sending literal "BEGIN"/"COMMIT" SQL over the wire, so
this check never issues those itself -- it only calls conn.commit()/
conn.rollback() and lets psycopg2 generate the SQL text, exercising the
same plain-SQL-forwarding behavior tests/pgclient already covers from Go.
"""
import sys

import psycopg2


def main():
    if len(sys.argv) != 2:
        print("usage: check.py <dsn>", file=sys.stderr)
        sys.exit(2)

    conn = psycopg2.connect(sys.argv[1])
    try:
        check(conn)
    finally:
        conn.close()
    print("OK")


def check(conn):
    cur = conn.cursor()

    cur.execute("SELECT 1")
    (one,) = cur.fetchone()
    assert one == 1, f"SELECT 1 returned {one!r}"

    cur.execute("SELECT 3.5")
    (f,) = cur.fetchone()
    assert f == 3.5, f"SELECT 3.5 returned {f!r}"

    cur.execute("SELECT 'hello'")
    (s,) = cur.fetchone()
    assert s == "hello", f"SELECT 'hello' returned {s!r}"

    cur.execute("SELECT x'00ff'")
    (blob,) = cur.fetchone()
    assert bytes(blob) == b"\x00\xff", f"SELECT x'00ff' returned {blob!r}"

    cur.execute("SELECT NULL")
    (null,) = cur.fetchone()
    assert null is None, f"SELECT NULL returned {null!r}"
    conn.commit()

    # spec §2: DDL must be rejected via the external I/F, and psycopg2 must
    # surface it as a structured error with SQLSTATE 42501 (spec §8), not
    # just some generic failure.
    try:
        cur.execute("CREATE TABLE psycopg_should_not_exist(a INTEGER)")
        conn.commit()
        raise AssertionError("expected CREATE TABLE to be rejected via the external I/F")
    except psycopg2.Error as e:
        conn.rollback()
        assert e.pgcode == "42501", f"expected SQLSTATE 42501, got {e.pgcode!r} ({e})"

    # A basic write/read round trip against the table run-all.sh seeded.
    cur.execute("INSERT INTO t VALUES (777001)")
    conn.commit()
    cur.execute("SELECT count(*) FROM t WHERE a = 777001")
    (n,) = cur.fetchone()
    assert n == 1, f"expected count=1 after INSERT+COMMIT, got {n}"


if __name__ == "__main__":
    main()
